Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/consent/one/DccTicketingConsentOneViewModel.kt

```diff
@@ -1,31 +1,138 @@
+import de.rki.coronawarnapp.dccticketing.ui.shared.DccTicketingSharedViewModel
 import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.encryption.ec.EcKeyGenerator
 import de.rki.coronawarnapp.util.ui.SingleLiveEvent
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
 import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import timber.log.Timber
 
 class DccTicketingConsentOneViewModel @AssistedInject constructor(
-    @Assisted private val dccTicketingTransactionContext: DccTicketingTransactionContext,
-    dispatcherProvider: DispatcherProvider
+    @Assisted private val dccTicketingSharedViewModel: DccTicketingSharedViewModel,
+    dispatcherProvider: DispatcherProvider,
+    private val dccTicketingRequestService: DccTicketingRequestService,
+    private val ecKeyGenerator: EcKeyGenerator
 ) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
 
-    val showCloseDialog = SingleLiveEvent<Unit>()
+    private val currentIsLoading = MutableStateFlow(false)
+    val isLoading: LiveData<Boolean> = currentIsLoading.asLiveData()
 
-    val uiState = MutableStateFlow(UiState(dccTicketingTransactionContext)).asLiveData()
+    private val currentEvent = SingleLiveEvent<DccTicketingConsentOneEvent>()
+    val events = currentEvent
 
-    fun goBack() {
-        showCloseDialog.postValue(Unit)
+    private val currentUiState = dccTicketingSharedViewModel.transactionContext
+        .map { UiState(it) }
+    val uiState: LiveData<UiState> = currentUiState.asLiveData2()
+
+    fun onUserCancel() {
+        Timber.d("onUserCancel()")
+        postEvent(ShowCancelConfirmationDialog)
+    }
+
+    fun goBack() = postEvent(NavigateBack)
+
+    fun showPrivacyInformation() = postEvent(NavigateToPrivacyInformation)
+
+    fun onUserConsent(): Unit = launch {
+        Timber.d("onUserConsent()")
+        currentIsLoading.compareAndSet(expect = false, update = true)
+        processUserConsent()
+        currentIsLoading.compareAndSet(expect = true, update = false)
+    }
+
+    private suspend fun processUserConsent() = try {
+        Timber.d("processUserConsent()")
+        dccTicketingSharedViewModel.apply {
+            transactionContext.first()
+                .requestServiceIdentityDocumentOfValidationService()
+                .generateECKeyPair()
+                .requestAccessToken()
+                .also { updateTransactionContext(ctx = it) }
+        }
+        postEvent(NavigateToCertificateSelection)
+    } catch (e: DccTicketingException) {
+        Timber.d(e, "Error while processing user consent")
+        val serviceProvider = currentUiState.first().provider
+        val lazyErrorMessage = e.errorMessage(serviceProvider = serviceProvider)
+        postEvent(ShowErrorDialog(lazyErrorMessage = lazyErrorMessage))
+    }
```

Comment:
I think we should handle `Exception` too, for example keyPair Generation can throw an exception in this case it is not DccTicketingException and it will crash the App. Better catch it and display error message